File: 1. Mayo/Códigos/Matlab/compare_performance.m
%% BENCHMARK MATLAB

% Warmup
warmup_gpu();

% Test con diferentes valores de N
test(5000);
test(10000);



function warmup_gpu()
    disp(" Warming up the GPU ");
    dim = 1000;
    matrix = gpuArray.rand(dim, dim, 'single');
    vector = gpuArray.rand(dim, 1, 'single');
    result = matrix * vector;
end

function test(dim)
    disp(" Comparing performance for dimension N = " + num2str(dim));

    % CPU Time
    np_matrix = rand(dim, dim, 'single');
    np_vector = rand(dim, 1, 'single');
    tic;
    for i = 1:10000
        numpy_result = i * np_matrix * np_vector;
    end
    numpy_time = toc;

    % GPU Time
    cp_matrix = gpuArray(rand(dim, dim, 'single'));
    cp_vector = gpuArray(rand(dim, 1, 'single'));
    tic;
    for i = 1:10000
        cupy_result = i * cp_matrix * cp_vector;
    end
    cupy_time = toc;

    % Transfer Time
    tic;
    np_cupy_result = gather(cupy_result);
    cupy_transfer_time = toc;

    disp("NumPy (CPU) Time: " + num2str(numpy_time) + " seconds");
    disp("CuPy (GPU) Time: " + num2str(cupy_time) + " seconds");
end

File: 1. Mayo/Códigos/Python/compare_performance/compare_performance/compare_performance.py
import numpy as np
import cupy as cp
import time
import os

os.environ["OMP_NUM_THREADS"] = "12" 


def numpy_random_matrix(dim, preallocated_matrix):
    preallocated_matrix[...] = np.random.rand(dim, dim).astype(np.float32)

def numpy_random_vector(dim, preallocated_vector):
    preallocated_vector[...] = np.random.rand(dim).astype(np.float32)

def cupy_random_matrix(dim, preallocated_matrix):
    preallocated_matrix[...] = cp.random.rand(dim, dim, dtype=cp.float32)

def cupy_random_vector(dim, preallocated_vector):
    preallocated_vector[...] = cp.random.rand(dim, dtype=cp.float32)

def numpy_dot(matrix, vector):
    return np.dot(matrix, vector)

def cupy_dot(matrix, vector):
    return cp.dot(matrix, vector)

def warmup_gpu():
    print("\nWarming up the GPU\n")
    dim = 1000
    matrix = cp.empty((dim, dim), dtype=cp.float32)
    vector = cp.empty(dim, dtype=cp.float32)
    cupy_random_matrix(dim, matrix)
    cupy_random_vector(dim, vector)
    result = cupy_dot(matrix, vector)




def parallel_dot(matrix, vector, iterations):
    results = Parallel(n_jobs=-1)(delayed(np.dot)(matrix, vector) for i in range(iterations))
    return results

def test_cpu(dim):
    print("\nComparing performance for dimension N = {}\n".format(dim))
    
    # CPU benchmark
    np_matrix = np.empty((dim, dim), dtype=np.float32)
    np_vector = np.empty(dim, dtype=np.float32)
    numpy_random_matrix(dim, np_matrix)
    numpy_random_vector(dim, np_vector)
    
    start = time.time()
    for i in range(100000):
         numpy_result = i * numpy_dot(np_matrix, np_vector)
    numpy_time = time.time() - start
    print("NumPy (CPU) Time: {:.6f} seconds".format(numpy_time))

    

   


def test_gpu(dim):
    print("\nComparing performance for dimension N = {}\n".format(dim))
    
    # Warmup
    warmup_gpu()
    

    # GPU benchmark
    cp_matrix = cp.empty((dim, dim), dtype=cp.float32)
    cp_vector = cp.empty(dim, dtype=cp.float32)
    cupy_random_matrix(dim, cp_matrix)
    cupy_random_vector(dim, cp_vector)
    
    start = time.time()
    for i in range(10000):
        cupy_result = i * cupy_dot(cp_matrix, cp_vector)
    cupy_time = time.time() - start
    print("CuPy (GPU) Time: {:.6f} seconds".format(cupy_time))



    





